### ISSUE TITLE
Request advice on serializing result-sets to JSON

### ISSUE THREAD

**Issue by phrrngtn** [nanodbc/nanodbc#356]
Hi, I am a database oriented programmer active in SQL and Python, mostly on SQL Server and SQLite and have some limited experience with ODBC via pyodbc. I am interested in embedding nanodbc within SQLite as an extension with a view to implementing some Foreign Data Wrapper capabilities within SQLite. Ultimately, I would like to map remote tables into local virtual tables. My initial goal is more modest and that is what I would like some help with here. 

What would be a good (low code) way of serializing an ODBC result-set to JSON using something like https://github.com/nlohmann/json ? I would like to use this technique to bootstrap my implementation with simple scalar functions that return JSON and post-process the result blob into a result-set using SQLite's built-in JSON processing capabilities. I have done very little C++ programming since the early 1990's so I think I have a fair idea of how to do this in a basic way e.g. retrieving the entire resultset, one array per column and serializing each column using techniques described in https://github.com/nlohmann/json#conversion-from-stl-containers and returning a header+body representation.

However, I am interested if you have any ideas on more slick/modern ways to do it, perhaps using iterators (or some other C++ mechanism that I don't know about)  and techniques from https://github.com/nlohmann/json#how-do-i-convert-third-party-types and return an array of dicts,  or header + array of arrays/dicts.

I have been hugely impressed with the power and performance of the JSON processing in SQLite (especially from 3.38 onwards)  and have been very pleasantly surprised by how easy some solutions work out by separating stabilization/persistence (i.e. getting the blob into the database) and normalization (shredding a subset of the data into regular tables with primary keys, foreign keys etc.)

Looking forward to suggestions. I hope that given the very narrow definition of the problem, I will be able to get away with learning a small subset of modern C++. I would like to avoid an outcome where I end up 'writing C in C++'. Note that I am not looking for any assistance wrt the SQLite C API side of things as I have up to date experience with that.

thanks in advance,
pjjH

## Environment

* OS and Compiler:
All OS;  C++11
I will do the development on WSL2 (Ubuntu) with g++ and unixODBC as the driver manager and SQL Server and PostgreSQL backends.

**Comment by phrrngtn** [comment]
I  have a working example with strings and can see how to use a  a switch statement on result.column_c_datatype to do `result.get<string>` `result.get<int>` etc. Rather than have to do the lookup for every column value in the result, it is possible to lookup the appropriate get method pointers for each column, stick them into an array and call them while processing the body of the results?

```cpp
json j;
for(int i=0; i< result.columns(); i++) {
       j[result.column_name(i)] = result.get<string>(i);
}
cout << j;
```

**Comment by mloskot** [comment]
The column type is run-time information per result-set, so it is necessary to resolve it for every result-set, for first row only is enough, then cache that (meta)information.
Reading into `std::variant` (still a feature request https://github.com/nanodbc/nanodbc/issues/324) or into Win32 `_variant_t` (implemented https://github.com/nanodbc/nanodbc/issues/327) may be a better/cleaner option.

**Comment by phrrngtn** [comment]
variant does indeed look cleaner. However, I will see what I can do with a the vector of pointers or vector of lambdas approach

Thanks!

**Comment by mloskot** [comment]
I'd go for plain switch as simplest to code and let/see if the compiler can optimise it for me, like here https://stackoverflow.com/a/30062376

**Comment by phrrngtn** [comment]
Good point. I will actually codgen that C++ switch in SQL!  The embedded templating processing in SQLite that I did a few weeks ago (https://github.com/phrrngtn/sqlite-template-inja)  is turning out to be very useful. In fact, the motivation to do the ODBC embedded in SQLite so that I can do catalog queries against remote, heterogenous databases and then run that through some DDL templates to generate some local, derived artifacts.

Thanks for your help. I have done a bit of low-level ODBC programming in the past and it is just great how much simpler the application code is with nanodbc.

**Comment by mloskot** [comment]
Sounds like an interesting and useful project indeed. Although I'm a bit swamped now, I'll try help with nanodbc issues.
